File: pysinthe/simulation.py
#! /usr/bin/env python3

import random
import math

from .feature import Polymerase
from .eventsignal import Signal
from .choices import weighted_choice

CELL_VOLUME = float(8e-16)


class SpeciesTracker():
    """
    Tracks species' copy numbers and maintains promoter-to-polymer and
    species-to-reaction maps to easily look up which polymers contain and given
    promoter and which reactions involve a given species. These maps are
    necessarry to cache propensities in the simulation.

    TODO: Move propensity cache into this class?
    """
    def __init__(self):
        self.species = {}  # species-level reactant counts
        self.promoter_polymer_map = {}  # Map of which polymers contain a given
        # promoter
        self.species_reaction_map = {}  # Map of which binding reaction
        # involves a given reactant
        self.propensity_signal = Signal()

    def increment_species(self, species_name, copy_number):
        """
        Change a species count by a given value (positive or negative).

        :param species_name: name of species to change count
        :param copy_number: increase current copy number by this number
        """
        if species_name in self.species:
            self.species[species_name] += copy_number
        else:
            self.species[species_name] = copy_number

        if copy_number == 0:
            return

        # Tell simulation to update propensity cache
        if species_name in self.species_reaction_map:
            for reaction in self.species_reaction_map[species_name]:
                self.propensity_signal.fire(reaction.index)

    def add_reaction(self, species_name, reaction):
        """
        Add a species-reaction pair to species-reaction map.

        :param species_name: name of species
        :param reaction: reaction object taht involves species
        """
        if species_name not in self.species_reaction_map:
            self.species_reaction_map[species_name] = [reaction]
        else:
            self.species_reaction_map[species_name].append(reaction)

    def add_polymer(self, promoter_name, polymer):
        """
        Add promoter-polymer pair to promoter-polymer map.

        :param promoter_name: name of promoter
        :param polymer: polymer that contains the named promoter
        """
        if promoter_name not in self.promoter_polymer_map:
            self.promoter_polymer_map[promoter_name] = [polymer]
        else:
            self.promoter_polymer_map[promoter_name].append(polymer)

    def polymers_with_promoter(self, promoter_name):
        """
        Get polymers that contain a given promoter.

        :param promoter_name: name of promoter

        :returns: list of polymer objects
        """
        return self.promoter_polymer_map[promoter_name]

    def reactions_with_species(self, species_name):
        """
        Get reactions that involve a given species.

        :param species_name: name of species

        :returns: list of reaction objects
        """
        return self.species_reaction_map[species_name]


class Reaction():
    """
    Generic class for a reaction. (Not currently used).
    """
    _AVAGADRO = float(6.0221409e+23)

    def __init__(self):
        self.index = -1  # Index inside propensity list, sim.alpha_list

    def calculate_propensity(self):
        """
        Calculate the propensity of this reaction
        """
        pass

    def execute(self):
        """
        Execute the reaction.
        """
        pass


class SpeciesReaction(Reaction):
    """
    Generic class for species-level reaction.
    """
    def __init__(self, tracker, rate_constant, reactants, products):
        super().__init__()
        self.tracker = tracker
        if len(reactants) > 2:
            raise RuntimeError("Simulation does not support reactions with "
                               "more than two reactant species.")
        if len(reactants) == 2:
            self.rate_constant = float(rate_constant)/(self._AVAGADRO*CELL_VOLUME)
        else:
            self.rate_constant = float(rate_constant)
        self.reactants = reactants
        self.products = products

        for reactant in self.reactants:
            self.tracker.add_reaction(reactant, self)
            self.tracker.increment_species(reactant, 0)

        for product in self.products:
            self.tracker.add_reaction(product, self)
            self.tracker.increment_species(product, 0)

    def calculate_propensity(self):
        """
        Calculate the propensity of this reaction
        """
        propensity = self.rate_constant
        for reactant in self.reactants:
            propensity *= self.tracker.species[reactant]

        return propensity

    def execute(self):
        """
        Execute the reaction.
        """
        for reactant in self.reactants:
            self.tracker.increment_species(reactant, -1)

        for product in self.products:
            self.tracker.increment_species(product, 1)


class Bind(Reaction):
    """
    Bind a polymerase to a polymer.
    """
    def __init__(self, tracker, rate_constant, promoter_name, pol_args):
        """

        TODO: Refactor so Bind inherits from SpeciesReaction

        :param sim: reference to simulation object in which this reaction occurs
        :param rate_constant: rate constant of reaction
        :param promoter_name: name of promoter involved in this reaction
        :param pol_args: list of arguments to pass to polymerase
            constructor upon execution of this reaction.
        """
        super().__init__()
        self.tracker = tracker
        self.polymerase = pol_args[0]
        self.pol_args = pol_args
        self.rate_constant = float(rate_constant)/(self._AVAGADRO*CELL_VOLUME)
        self.promoter_name = promoter_name

        self.tracker.add_reaction(self.promoter_name, self)
        self.tracker.add_reaction(self.polymerase, self)

    def calculate_propensity(self):
        """
        Calculate the propensity of this reaction.

        :returns: propensity of this reaction.
        """

        return self.rate_constant * self.tracker.species[self.polymerase]\
            * self.tracker.species[self.promoter_name]

    def execute(self):
        """
        Decrement reactants, choose a polymer, construct new polymerase, and
        bind the polymerase to the polymer.
        """
        # Find which polymer we should bind to
        weights = []
        for polymer in self.tracker.promoter_polymer_map[self.promoter_name]:
            # Weight by the number of unbound promoters in each polymer
            weights.append(polymer.count_uncovered(self.promoter_name))
        polymer = weighted_choice(
            self.tracker.promoter_polymer_map[self.promoter_name],
            weights=weights
        )
        # Construct new polymerase
        new_pol = Polymerase(*self.pol_args)
        polymer.bind_polymerase(new_pol, self.promoter_name)

        self.tracker.increment_species(self.promoter_name, -1)
        self.tracker.increment_species(self.polymerase, -1)

    def __str__(self):
        return self.promoter_name + "-" + self.polymerase


class Bridge(Reaction):
    """
    Encapsulate polymer so it can participate in species-level reaction
    processing.
    """
    def __init__(self, polymer):
        """
        :param polymer: polymer object that this reaction is encapsulating
        """
        super().__init__()
        self.polymer = polymer
        self.propensity_signal = Signal()  # Signal to fire when propensity
        # needs to be updated
        polymer.propensity_signal.connect(self.update)

    def calculate_propensity(self):
        """
        Retrieve the total propensity of all reactions that may occur within
        this polymer.

        :returns: total propensity of reactions within polymer
        """
        return self.polymer.calculate_propensity()

    def execute(self):
        """
        Execute reaction within polymer (e.g. typically moving a polymerase).
        """
        self.polymer.execute()

    def update(self):
        """
        Fires when polymer removes a polymerase from polymer and propensity
        needs to be updated.
        """
        self.propensity_signal.fire(self.index)

    def __str__(self):
        """
        Retrieve string representation of polymer.

        :returns: string representation of polymer encapsulated by this object
        """
        return str(self.polymer)


class Simulation:
    """
    Coordinate polymers and species-level reactions.
    """
    def __init__(self):
        self.time = 0  # simulation time
        self.runtime = 0
        self.time_step = 0
        self.debug = False
        self.terminations = {}  # track when polymerases terminate
        self.tracker = SpeciesTracker()
        self.reactions = []  # all reactions
        self.iteration = 0  # iteration counter
        self.alpha_list = []

        self.tracker.propensity_signal.connect(self.update_propensity)

    def run(self):
        print(self)
        old_time = 0
        while self.time < self.runtime:
            # Execute simulatio
            self.execute()
            if abs(self.time - old_time) > self.time_step:
                # Output data every ~time_step
                print(self)
                old_time += self.time_step
            elif self.debug is True:
                print(self)
                for pol in self.reactions:
                    print(pol)

    def register_reaction(self, reaction):
        """
        Add a SpeciesReaction object to the list of reactions.

        :param reaction: SpeciesReaction object
        """
        if reaction not in self.reactions:
            reaction.index = len(self.reactions)
            try:
                self.alpha_list.append(reaction.calculate_propensity())
            except KeyError:
                self.alpha_list.append(0.0)
            self.reactions.append(reaction)

    def register_polymer(self, polymer):
        """
        Add a polymer to the simulation.

        :param polymer: polymer object
        """
        # Add polymer to promoter-polymerase map
        for element in polymer.elements:
            if element.type == "promoter":
                self.tracker.add_polymer(element.name, polymer)

        # Encapsulate polymer in Bridge reaction and add to reaction list
        bridge = Bridge(polymer)
        bridge.propensity_signal.connect(self.update_propensity)
        self.register_reaction(bridge)

    def initialize_propensity(self):
        """
        Initialize all propensities before the start of the simulation.
        """
        for index, reaction in enumerate(self.reactions):
            self.alpha_list[index] = reaction.calculate_propensity()

        prop_sum = sum(self.alpha_list)

        if prop_sum == 0:
            raise ValueError("Initial global reaction propensity is 0. Make "
                             "sure that at least one promoter is exposed in "
                             "the genome.")

    def update_propensity(self, index):
        """
        Update a propensity of a reaction at a given index.
        """
        self.alpha_list[index] = float(
            self.reactions[index].calculate_propensity()
            )

    def execute(self):
        """
        Execute one cycle of reaction.
        """
        # Generate random number
        random_num = random.random()

        # Sum propensities
        alpha = sum(self.alpha_list)

        # Calculate tau, i.e. time until next reaction
        tau = (1/alpha)*math.log(1/random_num)
        self.time += tau

        # Randomly select next reaction to execute, weighted by propensities
        # print(self.reactions)
        next_reaction = weighted_choice(self.reactions,
                                        weights=self.alpha_list)
        next_reaction.execute()

        self.iteration += 1

    def free_promoter(self, species):
        """
        Increment promoter count at species level.
        """
        self.tracker.increment_species(species, 1)

    def block_promoter(self, species):
        """
        Decrement promoter count at species level.
        """
        self.tracker.increment_species(species, -1)

    def register_transcript(self, polymer):
        """
        Register a new transcript with the simulation.

        :param polymer: transcript object to be added to simulation
        :param reactants: proteins to be produced by this transcript
        """
        self.register_polymer(polymer)
        # Connect signals
        polymer.promoter_signal.connect(self.free_promoter)
        polymer.termination_signal.connect(self.terminate_translation)

    def terminate_transcription(self, species):
        """
        Terminate transcription.

        :param polymer: the newly constructed transcript
        :param species: name of the polymerase completing transcription
        :param reactants: list of reactants to add to species-level pool
            (usually RBSs)
        """
        self.tracker.increment_species(species, 1)
        self.count_termination("transcript")

    def terminate_translation(self, protein, species):
        """
        Terminate translation.

        :param name: name of the protein being produced
        :param species: name of species that just translated this protein
            (usually 'ribosome')
        """
        self.tracker.increment_species(species, 1)
        self.tracker.increment_species(protein, 1)
        self.count_termination(protein)

    def count_termination(self, name):
        """
        Record when a polymerase reaches a terminator.
        """
        name = name + "_total"
        if name not in self.terminations.keys():
            self.terminations[name] = 1
        else:
            self.terminations[name] += 1

    def __str__(self):
        """
        Convert `Simulation` to string representation showing names and
        termination times of `Polymerase` objects.
        """
        out_string = ""
        for name, count in self.terminations.items():
            out_string += str(self.iteration) + ", " + str(float(self.time)) + \
                ", " + name + ", " + str(count) + "\n"
        for name, count in self.tracker.species.items():
            out_string += str(self.iteration) + ", " + str(float(self.time)) + \
                ", " + name + ", " + str(count) + "\n"
        return out_string.strip()

File: pysinthe/simulation.cpp
#include <cmath>
#include <fstream>
#include <iostream>
#include <pybind11/pybind11.h>

#include "choices.hpp"
#include "polymer.hpp"
#include "simulation.hpp"

namespace py = pybind11;

const static double AVAGADRO = double(6.0221409e+23);
const static double CELL_VOLUME = double(8e-15);

SpeciesReaction::SpeciesReaction(double rate_constant,
                                 const std::vector<std::string> &reactants,
                                 const std::vector<std::string> &products)
    : rate_constant_(rate_constant), reactants_(reactants),
      products_(products) {
  // Error checking
  if (reactants_.size() > 2) {
    throw std::runtime_error("Simulation does not support reactions with "
                             "more than two reactant species.");
  }
  // Rate constant has to be transformed from macroscopic to mesoscopic
  if (reactants_.size() == 2) {
    rate_constant_ = rate_constant_ / (AVAGADRO * CELL_VOLUME);
  }
}

void SpeciesReaction::InitCounts() {
  auto &tracker = SpeciesTracker::Instance();
  for (const auto &reactant : reactants_) {
    tracker.Add(reactant, shared_from_this());
    tracker.Increment(reactant, 0);
  }
  for (const auto &product : products_) {
    tracker.Add(product, shared_from_this());
    tracker.Increment(product, 0);
  }
}

double SpeciesReaction::CalculatePropensity() {
  double propensity = rate_constant_;
  for (const auto &reactant : reactants_) {
    propensity *= SpeciesTracker::Instance().species(reactant);
  }
  return propensity;
}

void SpeciesReaction::Execute() {
  for (const auto &reactant : reactants_) {
    SpeciesTracker::Instance().Increment(reactant, -1);
  }
  for (const auto &product : products_) {
    SpeciesTracker::Instance().Increment(product, 1);
  }
}

Bind::Bind(double rate_constant, const std::string &promoter_name,
           const Polymerase &pol_template)
    : rate_constant_(rate_constant), promoter_name_(promoter_name),
      pol_name_(pol_template.name()), pol_template_(pol_template) {
  rate_constant_ = rate_constant_ / (AVAGADRO * CELL_VOLUME);
}

void Bind::InitCounts() {
  auto &tracker = SpeciesTracker::Instance();
  tracker.Add(promoter_name_, shared_from_this());
  tracker.Add(pol_name_, shared_from_this());
}

double Bind::CalculatePropensity() {
  auto &tracker = SpeciesTracker::Instance();
  return rate_constant_ * tracker.species(pol_name_) *
         tracker.species(promoter_name_);
}

void Bind::Execute() {
  // std::cout << "binding..." + std::to_string(CalculatePropensity())
  //           << std::endl;
  auto &tracker = SpeciesTracker::Instance();
  auto weights = std::vector<double>();
  for (const auto &polymer : tracker.FindPolymers(promoter_name_)) {
    weights.push_back(double(polymer->uncovered(promoter_name_)));
  }
  Polymer::Ptr polymer =
      Random::WeightedChoice(tracker.FindPolymers(promoter_name_), weights);
  auto new_pol = std::make_shared<Polymerase>(pol_template_);
  polymer->Bind(new_pol, promoter_name_);
  tracker.propensity_signal_.Emit(polymer->index());
  tracker.Increment(promoter_name_, -1);
  tracker.Increment(pol_name_, -1);
}

Simulation::Simulation()
    : time_(0), stop_time_(0), time_step_(0), alpha_sum_(0) {
  auto &tracker = SpeciesTracker::Instance();
  tracker.Clear();
}

void Simulation::Run(const std::string &output_name) {
  auto &tracker = SpeciesTracker::Instance();
  if (time_ == 0) {
    tracker.propensity_signal_.ConnectMember(shared_from_this(),
                                             &Simulation::UpdatePropensity);
    InitPropensity();
  }
  // Set up file output streams
  std::ofstream countfile(output_name + "_counts.tsv", std::ios::trunc);
  int out_time = 0;
  while (time_ < stop_time_) {
    Execute();
    if ((out_time - time_) < 0.001) {
      for (auto elem : tracker.species()) {
        countfile << (std::to_string(time_) + "\t" + elem.first + "\t" +
                      std::to_string(elem.second))
                  << std::endl;
      }
      countfile.flush();
      out_time += time_step_;
    }
  }
  countfile.close();
}

void Simulation::RegisterReaction(Reaction::Ptr reaction) {
  reaction->InitCounts();
  auto it = std::find(reactions_.begin(), reactions_.end(), reaction);
  if (it == reactions_.end()) {
    reaction->index(reactions_.size());
    double new_prop = reaction->CalculatePropensity();
    alpha_list_.push_back(new_prop);
    alpha_sum_ += new_prop;
    reactions_.push_back(reaction);
  }
}

void Simulation::RegisterPolymer(Polymer::Ptr polymer) {
  polymer->InitElements();
  for (auto &elem : polymer->elements()) {
    if (elem->type() == "promoter") {
      elem->uncover_signal_.ConnectMember(shared_from_this(),
                                          &Simulation::FreePromoter);
      elem->cover_signal_.ConnectMember(shared_from_this(),
                                        &Simulation::BlockPromoter);
      SpeciesTracker::Instance().Add(elem->name(), polymer);
    }
  }
  // Encapsulate polymer in Bridge reaction and add to reaction list
  auto bridge = std::make_shared<Bridge>(polymer);
  RegisterReaction(bridge);
  polymer->index(bridge->index());
}

void Simulation::RegisterGenome(Genome::Ptr genome) {
  RegisterPolymer(genome);
  genome->termination_signal_.ConnectMember(
      shared_from_this(), &Simulation::TerminateTranscription);
  genome->transcript_signal_.ConnectMember(shared_from_this(),
                                           &Simulation::RegisterTranscript);
}

void Simulation::RegisterTranscript(Transcript::Ptr transcript) {
  RegisterPolymer(transcript);
  transcript->termination_signal_.ConnectMember(
      shared_from_this(), &Simulation::TerminateTranslation);
}

void Simulation::InitPropensity() {
  for (int i = 0; i < reactions_.size(); i++) {
    UpdatePropensity(i);
  }
  // Make sure prop sum is starting from 0
  alpha_sum_ = 0;
  for (const auto &alpha : alpha_list_) {
    alpha_sum_ += alpha;
  }
}

void Simulation::UpdatePropensity(int index) {
  double new_prop = reactions_[index]->CalculatePropensity();
  // std::cout << "upating prop..." + std::to_string(new_prop) << std::endl;
  double diff = new_prop - alpha_list_[index];
  alpha_sum_ += diff;
  alpha_list_[index] = new_prop;
}

void Simulation::Execute() {
  // Generate random number
  if (alpha_sum_ == 0) {
    throw std::runtime_error("Propensity of system is 0.");
  }
  // InitPropensity();
  double random_num = Random::random();
  // Calculate tau, i.e. time until next reaction
  double tau = (1.0 / alpha_sum_) * std::log(1.0 / random_num);
  time_ += tau;
  // for (auto i : alpha_list_) {
  //   std::cout << std::to_string(i) + " " << std::endl;
  // }
  // Randomly select next reaction to execute, weighted by propensities
  auto next_reaction = Random::WeightedChoice(reactions_, alpha_list_);
  // std::cout << "attempting to bind..." +
  //                  std::to_string(alpha_list_[next_reaction->index()])
  //           << std::endl;
  next_reaction->Execute();
  iteration_++;
}

void Simulation::FreePromoter(const std::string &species_name) {
  SpeciesTracker::Instance().Increment(species_name, 1);
}
void Simulation::BlockPromoter(const std::string &species_name) {
  SpeciesTracker::Instance().Increment(species_name, -1);
}

void Simulation::TerminateTranscription(int polymer_index,
                                        const std::string &pol_name,
                                        const std::string &gene_name) {
  SpeciesTracker::Instance().Increment(pol_name, 1);
  UpdatePropensity(polymer_index);
  CountTermination("transcript");
}

void Simulation::TerminateTranslation(int polymer_index,
                                      const std::string &pol_name,
                                      const std::string &gene_name) {
  SpeciesTracker::Instance().Increment(pol_name, 1);
  SpeciesTracker::Instance().Increment(gene_name, 1);
  UpdatePropensity(polymer_index);
  CountTermination(gene_name);
}

void Simulation::CountTermination(const std::string &name) {
  auto new_name = name + "_total";
  if (terminations_.count(name) == 0) {
    terminations_[new_name] = 1;
  } else {
    terminations_[new_name]++;
  }
}

SpeciesTracker &SpeciesTracker::Instance() {
  static SpeciesTracker instance;
  return instance;
}

void SpeciesTracker::Clear() {
  species_.clear();
  promoter_map_.clear();
  species_map_.clear();
  propensity_signal_ = Signal<int>();
}

void SpeciesTracker::Register(SpeciesReaction::Ptr reaction) {
  // Add this reaction to SpeciesTracker
  for (const auto &reactant : reaction->reactants()) {
    Add(reactant, reaction);
    // Initialize reactant counts (this usually gets set to some non-zero
    // value later)
    Increment(reactant, 0);
  }
  // Do the same for products
  for (const auto &product : reaction->products()) {
    Add(product, reaction);
    Increment(product, 0);
  }
}

void SpeciesTracker::Increment(const std::string &species_name,
                               int copy_number) {
  if (species_.count(species_name) == 0) {
    species_[species_name] = copy_number;
  } else {
    species_[species_name] += copy_number;
  }
  if (species_map_.count(species_name) > 0) {
    for (const auto &reaction : species_map_[species_name]) {
      propensity_signal_.Emit(reaction->index());
    }
  }
  if (species_[species_name] < 0) {
    throw std::runtime_error("Species count less than 0." + species_name);
  }
}

void SpeciesTracker::Add(const std::string &species_name,
                         Reaction::Ptr reaction) {
  if (species_map_.count(species_name) == 0) {
    species_map_[species_name] = Reaction::VecPtr{reaction};
  } else {
    // TODO: Maybe use a better data type here like a set?
    auto it = std::find(species_map_[species_name].begin(),
                        species_map_[species_name].end(), reaction);
    if (it == species_map_[species_name].end()) {
      species_map_[species_name].push_back(reaction);
    }
  }
}

void SpeciesTracker::Add(const std::string &promoter_name,
                         Polymer::Ptr polymer) {
  if (promoter_map_.count(promoter_name) == 0) {
    promoter_map_[promoter_name] = Polymer::VecPtr{polymer};
  } else {
    promoter_map_[promoter_name].push_back(polymer);
  }
}

const Reaction::VecPtr &
SpeciesTracker::FindReactions(const std::string &species_name) {
  return species_map_[species_name];
}

const Polymer::VecPtr &
SpeciesTracker::FindPolymers(const std::string &promoter_name) {
  return promoter_map_[promoter_name];
}

int SpeciesTracker::species(const std::string &reactant) {
  return species_[reactant];
}